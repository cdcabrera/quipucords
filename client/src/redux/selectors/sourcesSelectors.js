import { createSelector } from 'reselect';
import _get from 'lodash/get';
import apiTypes from '../../constants/apiConstants';

/**
 * Map a source object to consumable prop names
 */
const sourceDetail = state => state.addSourceWizard.source;

const sourceDetailSelector = createSelector(
  [sourceDetail],
  source => {
    const updatedSource = {
      credentials: []
    };

    if (source) {
      /**
       * Allow for an edit source with credentials in the form of
       * [{ id:Number, type:String, name:String }]
       * Or an original/edited source in the form of
       * [id]
       */
      updatedSource.credentials = _get(source, apiTypes.API_RESPONSE_SOURCE_CREDENTIALS, []).map(
        cred => cred[apiTypes.API_RESPONSE_SOURCE_CREDENTIALS_ID] || cred
      );

      updatedSource.hosts = source[apiTypes.API_RESPONSE_SOURCE_HOSTS];
      updatedSource.id = source[apiTypes.API_RESPONSE_SOURCE_ID];
      updatedSource.name = source[apiTypes.API_RESPONSE_SOURCE_NAME];

      updatedSource.optionSslCert = _get(
        source[apiTypes.API_RESPONSE_SOURCE_OPTIONS],
        apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_CERT,
        false
      );

      updatedSource.optionSslProtocol = _get(
        source[apiTypes.API_RESPONSE_SOURCE_OPTIONS],
        apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_PROTOCOL,
        false
      );

      updatedSource.optionDisableSsl = _get(
        source[apiTypes.API_RESPONSE_SOURCE_OPTIONS],
        apiTypes.API_RESPONSE_SOURCE_OPTIONS_DISABLE_SSL,
        false
      );

      updatedSource.optionParmiko = _get(
        source[apiTypes.API_RESPONSE_SOURCE_OPTIONS],
        apiTypes.API_RESPONSE_SOURCE_OPTIONS_PARAMIKO,
        false
      );

      updatedSource.port = source[apiTypes.API_RESPONSE_SOURCE_PORT];
      updatedSource.type = source[apiTypes.API_RESPONSE_SOURCE_SOURCE_TYPE];
    }

    return updatedSource;
  }
);

const makeSourceDetailSelector = () => sourceDetailSelector;

const sourcesSelectors = {
  sourceDetail: sourceDetailSelector,
  makeSourceDetail: makeSourceDetailSelector
};

export { sourcesSelectors as default, sourcesSelectors, sourceDetailSelector, makeSourceDetailSelector };
