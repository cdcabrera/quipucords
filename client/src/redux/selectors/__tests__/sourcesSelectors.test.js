import sourcesSelectors from '../sourcesSelectors';
import apiTypes from '../../../constants/apiConstants';

describe('SourcesSelectors', () => {
  it('Should return specific selectors', () => {
    expect(sourcesSelectors).toMatchSnapshot('selectors');
  });

  it('Should map a source to consumable props', () => {
    const state = {
      addSourceWizard: {
        source: {
          [apiTypes.API_RESPONSE_SOURCE_CREDENTIALS]: [15],
          [apiTypes.API_RESPONSE_SOURCE_HOSTS]: ['192.168.0.1'],
          [apiTypes.API_RESPONSE_SOURCE_ID]: 1,
          [apiTypes.API_RESPONSE_SOURCE_NAME]: 'lorem',
          [apiTypes.API_RESPONSE_SOURCE_OPTIONS]: {
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_PARAMIKO]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_DISABLE_SSL]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_PROTOCOL]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_CERT]: false
          },
          [apiTypes.API_RESPONSE_SOURCE_PORT]: 22,
          [apiTypes.API_RESPONSE_SOURCE_SOURCE_TYPE]: 'network'
        }
      }
    };

    expect(sourcesSelectors.sourceDetail(state)).toMatchSnapshot('sourceDetail');

    const updatedState = {
      addSourceWizard: {
        source: {
          [apiTypes.API_RESPONSE_SOURCE_CREDENTIALS]: [
            {
              [apiTypes.API_RESPONSE_SOURCE_CREDENTIALS_ID]: 10
            }
          ],
          [apiTypes.API_RESPONSE_SOURCE_HOSTS]: ['192.168.0.1'],
          [apiTypes.API_RESPONSE_SOURCE_ID]: 1,
          [apiTypes.API_RESPONSE_SOURCE_NAME]: 'lorem',
          [apiTypes.API_RESPONSE_SOURCE_OPTIONS]: {
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_PARAMIKO]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_DISABLE_SSL]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_PROTOCOL]: false,
            [apiTypes.API_RESPONSE_SOURCE_OPTIONS_SSL_CERT]: false
          },
          [apiTypes.API_RESPONSE_SOURCE_PORT]: 22,
          [apiTypes.API_RESPONSE_SOURCE_SOURCE_TYPE]: 'network'
        }
      }
    };

    expect(sourcesSelectors.sourceDetail(updatedState)).toMatchSnapshot('sourceDetail updated');
  });

  it('Should map an undefined source', () => {
    const state = {
      addSourceWizard: {}
    };

    expect(sourcesSelectors.sourceDetail(state)).toMatchSnapshot('sourceDetail undefined');
  });
});
