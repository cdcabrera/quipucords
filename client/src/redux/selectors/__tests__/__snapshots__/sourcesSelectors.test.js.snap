// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors 1`] = `
Object {
  "makeSourceDetail": [Function],
  "sourceDetail": [Function],
}
`;

exports[`sourceDetail 1`] = `
Object {
  "credentials": Array [
    15,
  ],
  "hosts": Array [
    "192.168.0.1",
  ],
  "id": 1,
  "name": "lorem",
  "optionDisableSsl": false,
  "optionParmiko": false,
  "optionSslCert": false,
  "optionSslProtocol": false,
  "port": 22,
  "type": "network",
}
`;

exports[`sourceDetail undefined 1`] = `
Object {
  "credentials": Array [],
}
`;

exports[`sourceDetail updated 1`] = `
Object {
  "credentials": Array [
    10,
  ],
  "hosts": Array [
    "192.168.0.1",
  ],
  "id": 1,
  "name": "lorem",
  "optionDisableSsl": false,
  "optionParmiko": false,
  "optionSslCert": false,
  "optionSslProtocol": false,
  "port": 22,
  "type": "network",
}
`;
